01-resources-v2/proposal.md

```diff
@@ -0,0 +1,413 @@
+
+* Extend the versioned artifacts interface to support deletion of versions.
+
+
+# Proposal
+
+## General Types
+
+```go
+// Space is a name of a space, e.g. "master", "release/3.14", "1.0".
+type Space string
+
+// Config is a black box containing all user-supplied configuration, combining
+// `source` in the resource definition with `params` from the step (in the
+// case of `get` or `put`).
+type Config map[string]interface{}
+
+// Version is a key-value identifier for a version of a resource, e.g.
+// `{"ref":"abcdef"}`, `{"version":"1.2.3"}`.
+type Version map[string]string
+
+// Metadata is an ordered list of metadata fields to display to the user about
+// a resource version. It's ordered so that the resource can decide the best
+// way to show it.
+type Metadata []MetadataField
+
+// MetadataField is an arbitrary key-value to display to the user about a
+// version of a resource.
+type MetadataField struct {
+  Name  string `json:"name"`
+  Value string `json:"value"`
+}
+```
+
+## Versioned Artifacts interface
+
+### `check`: Detect versions across spaces.
+
+The `check` command will be invoked with the following JSON structure on
+`stdin`:
+
+```go
+// CheckRequest contains the resource's configuration and latest version
+// associated to each space.
+type CheckRequest struct {
+  Config       Config            `json:"config"`
+  From         map[Space]Version `json:"from"`
+  ResponsePath string            `json:"response_path"`
+}
+```
+
+The first call will have an empty object as `from`.
+
+Any spaces discovered by the resource but not present in `from` should collect
+versions from the beginning.
+
+For each space in `from`, the resource should collect all versions that appear
+*after* the current version, including the given version if it's still present.
+If the given version is no longer present, the resource should instead collect
+from the beginning, as if the space was not specified.
+
+If any space in `from` is no longer present, the resource should ignore it, and
+not include it in the response.
+
+The resource may limit the number of versions it returns however it likes in
+order to save RAM and incrementally build the history. The `has_latest` field in
+the response (see below) is used to indicate this scenario.
+
+The resource should also determine a "default space", if any. Having a default
+space is useful for things like Git repos which have a default branch, or
+version spaces (e.g. `1.8`, `2.0`) which can point to the latest version line by
+default. If there is no default space, the user must specify it explicitly in
+the pipeline, either by configuring one on the resource (`space: foo`) or on the
+`get` step (`spaces: [foo]`).
+
+The command should then emit the collected versions and default space (if any)
+as following JSON response structure the file specified by `response_path`:
+
+```go
+// CheckResponse returns the detected spaces and the default space, if any.
+type CheckResponse struct {
```

Comment:
One last thing I'm considering before closing out this proposal: should we write one big JSON response body, or have a more stream-based response interface?

For example, the response could look something like:

```js
{"spaces":["a","b","c"],"default_space":"a"}
{"space":"a","version":{"v":"1"},"metadata":[]}
{"space":"a","version":{"v":"2"},"metadata":[]}
{"space":"a","version":{"v":"3"},"metadata":[]}
{"space":"b","version":{"v":"1"},"metadata":[]}
{"space":"b","version":{"v":"2"},"metadata":[]}
{"space":"b","version":{"v":"3"},"metadata":[]}
// ...
```

So there'd be an initial JSON object returning the spaces and default space, and from then on it'd be a stream of versions associated to each space. This would enable resource authors to write versions as they find them rather than hold them all in memory and do one big JSON marshal. With this design could get rid of `has_latest` - disk space would now be the "at risk" resource, but it should be a lot cheaper than RAM.

The interface itself is a bit more complicated but I think it'd actually make it harder for resource authors to make mistakes. Authors would notice a broken implementation of a more complicated interface early, rather than implement the easier interface but then have emergent behavior like using a ton of RAM on large repos.

@itsdalmo @cwlbraa @topherbullock whaddaya think?

Reply:
Sorry for the late reply!

I'm wondering if it would be possible to have Concourse discover the possible spaces from the set of spaces as it reads this records? E.g., in the example below it would read one record at a time and keep a set in memory which ends up as (`["a", "b"]`):

```json
{"space":"a","version":{"v":"1"},"metadata":[]}
{"space":"b","version":{"v":"1"},"metadata":[]}
```

The reason I ask is because in e.g. a `github-pullrequest-resource` which uses the GraphQL API, it's possible (and desirable because it saves API calls) to just list all open PR's directly, irrespective of which branch or base it has. So in that case, the resource itself could not write the complete set of spaces (the first record in your example) until it had retrieved every open PR from the API, stored them in memory and then looped over them to create a set of unique spaces (e.g. base or PR branch).

Also, I'm a bit unsure about what `default_space` means in this context. Assuming that it would e.g. be the "default branch" of a Git repository, I'm not sure there are many other use cases where a `default_space` can be determined?

Reply:
@itsdalmo I think that's cleaner, but the one disadvantage is that it won't be able to return spaces that have no versions. Though perhaps that's not very useful anyway.

The notion of `default_space` would be used for two things:

1. In the `git` resource (probably any VCS resource), this would return the default branch.
1. For semver'd resources (e.g. Docker images, S3 artifacts, ...) this would return the space corresponding to the latest version. This is to match today's behavior where by default you may want to just "always roll forward".